src/main/java/hudson/plugins/git/GitSCM.java

```diff
@@ -1762,26 +1791,105 @@ public BuildData copyBuildData(Run build) {
-     * may not be recorded if an exception occurs in the plugin logic.
+     * Generate the build log (BuildData) from recorded BuildDetails.
      *
-     * @param build run whose build data is returned
-     * @return the last recorded build data
+     * @param build run whose build data should be generated.
+     * @return build data generated from historical build details
      */
     public @CheckForNull BuildData getBuildData(Run build) {
         BuildData buildData = null;
+
         while (build != null) {
-            List<BuildData> buildDataList = build.getActions(BuildData.class);
-            for (BuildData bd : buildDataList) {
-                if (bd != null && isRelevantBuildData(bd)) {
-                    buildData = bd;
-                    break;
+            BuildData oldBuildData = findRelevantBuildData(build);
+            if (oldBuildData != null) {
+                /* This is an older build which has BuildData saved directly.
+                 * If we haven't even started constructing a BuildData
+                 * structure yet, then the contents of this old BuildData
+                 * should be sufficient, and ensures we maintain compatibility
+                 * with older build history.
+                 */
+                if (buildData == null) {
+                    return oldBuildData;
+                }
+
+                /* Otherwise, we've found newer builds with valid BuildDetails,
+                 * so we'll just fill in any missing branches from this
+                 * BuildData first. We can stop digging further since we only
+                 * need a single BuildData to complete the branch names map.
+                 */
+                for (Build entry : oldBuildData.buildsByBranchName.values()) {
+                    addBuildByBranchNames(buildData.buildsByBranchName, entry);
+                    return buildData;
                 }
             }
-            if (buildData != null) {
-                break;
+
+            BuildDetails oldBuildDetails = findRelevantBuildDetails(build);
+            if (oldBuildDetails != null) {
+                if (buildData == null) {
+                    /* This is the first relevant BuildDetails we found, so we
+                     * need to create a new BuildData structure
+                     */
+                    buildData = new BuildData(oldBuildDetails);
+                } else {
+                    /* Otherwise, simply add this BuildDetails data into the
+                     * BuildData structure we're generating. Since this is
+                     * a top down construction, we will only add branches
+                     * that weren't build by "newer" builds already.
+                     */
+                    addBuildByBranchNames(buildData.buildsByBranchName, oldBuildDetails.build);
+                }
+
             }
+
+            /* Keep digging through build history until we run out or
```

Comment:
Not sure what the previous behavior was—under which circumstances the loop ran, and for how long—but any loop like this is a red flag in Jenkins as it is breaking the lazy-loading system introduced (IIRC) in 1.484. If you have a project with thousands of build records saved, reading each `build.xml` from disk and then unmarshalling its contents into Java objects can be quite expensive. (RAM usage balloons, and eventually `SoftReference`s get thrown out, at which point the _next_ such loop does it all over again, basically making the server grind to a halt.) Therefore we make every effort to avoid unbounded iteration of build records, at least in performance-critical points—i.e., any time during HTTP rendering or routine build processes.

Now if this code is only run once per historical build (or project?) and caches the relevant information along the way so that it need not loop the next time, that is OK.

Reply:
So, the previous behavior was that each build stored its own copy of the entire build data structure. This caused significant problem because *every* build.xml contained the entire build data history, which balloons if you have unique branches per build. (such as with gerrit).

Reply:
> Now if this code is only run once per historical build (or project?) and caches the relevant information along the way so that it need not loop the next time, that is OK.

This is definitely run once per build. I did test it using a server that generated thousands of builds, and didn't see any immediate problems. It's definitely not the most ideal solution, so better suggestions would be welcome.

I opted to use this method because it allowed us to keep feature parity with the older system and seemed to work ok.

Dropping the build data entirely is another option, but we lose the features it provides. Maybe those aren't worth it.... but I wasn't the one to make that call.

Reply:
More comments on this implementation vs the older one.

So *every* build has to load the BuildData somehow.

In the old implementation, we deserialize it from the build.xml of the previous build, *and* we update, copy and store it in the new build.xml of the next build. If you have unique branches, then this will continue to grow for every build, plus there is no mechanism to clean it out.

In the new method, we re-build the data by checking each previous build.xml for build details. Either way the same amount of data (or less, see below) must be read in the new way as in the old way. The old way *still* loads the data into memory and thus still bloats memory and takes time to garbage collect.

The new way has a few advantages:

1) we don't copy the build data and store outdated copies for every build.mxl
2) we don't keep infinite build data that can grow arbitrarily large, because we only rebuild based on the builds which have history kept. The old method build data could keep growing unbounded when using something like gerrit refs for branches, and it would only ever clean up manually.

The only difference I can tell, is that now we read build details from multiple build.xml files, (one per historical job), vs reading the data deserialized from one build.xml. However, the sizes are going to be the same *or smaller* in the new method. The one disadvantage is that in the old world of a single branch, the build data would not grow out of bounds, so it would be trivial to keep track of, and it is definitely more efficient, since we'd have to keep building back in time.

As far as I know though, we don't generate build data except during choosing what to build, so it shouldn't be a thing called back from the webpage, but only during git operations.

I think using a separate xml file could work as well, but is an even further significant refactor. Additionally, so is simply dropping the build data entirely, as there are potentially features which depend on it. I haven't figured the scope of what we'd lose, and I'm not the plugin maintainer who could make that decision.

In my tests, I had a jenkins server which ran with several thousand jobs as the history and it didn't appear to cause significant problems.

I'm fairly confident that this method is better than build data was, even if it's not the perfect solution.